File: dataStructure/stack.cpp
#include <cstdio>
#include <iostream>

using namespace std;

template <typename T>
class Stack {
 public:
  Stack(unsigned int len) : length(len), topIdx(-1) {
    items = new T[length];
  }
  ~Stack() {
    delete [] items;
  }

  void push(T item) {
    if (isFull()) cerr << "Stack is full!\n";
    else {
      topIdx++;
      items[topIdx] = item;
    }
  }

  void pop() {
    if (isEmpty()) cerr << "Stack is empty!\n";
    else topIdx--;
  }

  int size() {
    return topIdx + 1;
  }

  T top() {
    if (isEmpty()) cerr << "Stack is empty!\n";
    return items[topIdx];
  }

  bool isEmpty() {
    return topIdx == -1;
  }

  bool isFull() {
    return topIdx == length - 1;
  }

  void display() {
    printf("[");
    for (int i = topIdx; i >= 0; i--) cout << ' ' << items[i];
    printf(" ]\n");
  }


 private:
  int topIdx;
  unsigned int length;
  T *items;
};

int main() {
  Stack arr = Stack<int>(5);

  arr.push(1);
  arr.push(2);
  arr.push(3);
  arr.display();
  arr.pop();
  arr.push(4);
  arr.display();
  printf("%d\n", arr.top());

  return 0;
}

File: dataStructure/Stack.java
import java.io.PrintStream;

class Stack<T> {
  // `static final` defines a constant value
  static final PrintStream stdout = System.out;
  static final PrintStream stderr = System.err;

  private int top;
  private int length;
  private T[] items;

  public Stack(int length) {
    this.length = length;
    this.top = -1;
    this.items = (T[]) new Object[length];
  }

  public void push(T item) {
    if (this.isFull()) stderr.println("Stack is full!");
    else {
      this.top++;
      this.items[this.top] = item;
    }
  }

  public void pop() {
    if (this.isEmpty()) stderr.println("Stack is empty!");
    else this.top--;
  }

  public int size() {
    return this.top + 1;
  }

  public int length() {
    return this.length;
  }

  public T top() throws Error {
    if (this.isEmpty()) throw new Error("Stack is empty!");
    return this.items[this.top];
  }

  public boolean isEmpty() {
    return this.top == -1;
  }

  public boolean isFull() {
    return this.top == this.length - 1;
  }

  public void display() {
    stdout.printf("[");
    for (int i = this.top; i >= 0; i--) stdout.print(" " + this.items[i]);
    stdout.println(" ]");
  }
}
